### ISSUE TITLE
Manjaro kernels and nvidia drivers

### ISSUE THREAD

**Issue by ryanmusante** [Tk-Glitch/PKGBUILDS#13]
More a question than an issue. I apologize for my ignorance. `linux418-tkg, nvidia-dev-all, wine-tkg-git` seems like a perfect trio for dxvk/vulkan gaming. I'm already running wine/winetricks tkg and sometimes compile dxvk.

On manjaro-testing and was curious if kernel/nvidia-dev might cause problems with udev settings or detected modules in my configuration? 

From a mostly stock install, would this kernel be installed alongside the other and accessible from GRUB the same? Does mkinitcpio.conf affect these?

**Comment by Tk-Glitch** [comment]
I actually am running manjaro on my laptop with these packages installed. Neither the kernel nor the nvidia-dev drivers will cause problems with manjaro, and yes the kernel can be installed alongside others, and will be accessible from GRUB.

You'll "bypass" mhwd but unlike some other nvidia packages on AUR, it won't actually break it. If you happen to need to go back to mhwd nvidia install, just uninstall the tkg package and proceed with mhwd as usual. You may need to run **mkinitcpio -P** after updating your drivers, but you'll see a warning about it after install anyway.

On the kernel side, the optimal setup requires some kind of initial work so modprobed-db grabs everything correctly. It might not be needed depending on your setup, but better safe than sorry, right? :D If you're not planning to use modprobed-db, you can ignore what follows.

So you'll need to build a first tkg kernel that will be used to populate the modprobed-db (any flavour will do). You should NOT select to use modprobed-db for this one, of course.
It can take a while to build depending on your hardware.
Once this is done and installed, just run `update-grub` and you'll see your new kernel in GRUB.
If you didn't do it previously, follow the modprobed-db advices on the Arch-wiki.

Once you're pretty confident your db is complete, you can now compile your streamlined kernel by selecting to use modprobed-db this time. It'll build **much** faster than the previous one, and will have a **much** lower footprint. Also, you won't ever need to rebuild that "initial kernel" as long as you keep your db, and all subsequent compilations will be fast (and even faster in fact as it'll use ccache).

**Comment by ryanmusante** [comment]
I'm thinking it's about time I take my _~/.ccache_ off of tmpfs. I put it there but I realize that between boots, it makes it lose its purpose essentially. I even had _~/.nv_ for a time but same problem with that. 

I see that _modprobe-db_ has a systemd process that can be run, that sounds like the options I'll use in case of adds/removes. 

According to wiki, does your pkgbuild automate the calling of _localmodconfig_ ? Meaning as long as I start the service and have it automate that it won't "pollute" the kernel space?

> While using sudo is recommended, it is not necessary as the same effect can be achieved with make LSMOD=$HOME/.config/modprobed.db localmodconfig, this will also avoid polluting your kernel space with unnecessary modules.

Will successive updates from manjaro pull in nvidia standard drivers requiring me to rebuild locally from the pkgbuild?

**Comment by Tk-Glitch** [comment]
The pkgbuild will indeed automate the localmodconfig call if you answer "y" when you're asked if you want to use modprobed-db.

The service will only update the db regularly, it is more QoL than a requirement. You could update the db by hand and end up with the same result, basically. It won't "pollute" the kernel space as it's indeed called with `make LSMOD=$HOME/.config/modprobed.db localmodconfig` in the pkgbuild.

Manjaro updates won't replace the nvidia tkg packages as they don't have the same pkgname even though they are conflicting (and you'll be prompted to replace your currently installed nvidia packages with them). You'll get mhwd updates, but not the actual drivers.

By installing the dkms version of the nvidia tkg packages, the modules will be automagically rebuilt when you're updating your kernel(s) so you don't have to worry about rebuilding.
Special cases : GCC and xorg "major" updates can sometimes require a rebuild of the nvidia packages.

**Comment by ryanmusante** [comment]
I use mq-deadline on both of my ssds with the following tunables:
```
ATTR{queue/scheduler}="mq-deadline"
ATTR{queue/iosched/front_merges}="0"
ATTR{queue/iosched/read_expire}="500"
ATTR{queue/iosched/write_expire}="500"
ATTR{queue/iosched/fifo_batch}="1"
ATTR{queue/iosched/writes_starved}="2"
```
Which scheduler is closest to it? Will it automatically switch my scheduler to PDS or keep the settings I have in /etc/udev/rules.d?
```
What CPU sched variant do you want to build/install?
> 1.PDS
  2.MUQSS
  3.CFS
```

**Comment by Tk-Glitch** [comment]
`I put some questions up, I hope this isn't detracting you from your work. Thanks again for the support!`
Don't mention it!

`Why is modprobed-db necessary if lsmod lists all running modules? Should ~/.config/modprobed.db contain almost the same number as lsmod > list.txt?`
All running modules aren't necessarily (and, in fact, only rarely are) all the modules you'll end up needing. You didn't connect your USB drive? The module won't be loaded. You didn't mount an iso? The module won't be loaded. You didn't switch your CPU governor? Same story. That's why it's important to populate the db. You need to actually use everything on your system that could need to load a module or you'll be missing some.

`PDS and MUQSS are not the same as I/O schedulers? It isn't selectable after reboot.`
PDS and MuQSS, like CFS, are CPU/process schedulers. Since it's impacting a large part of the way the kernel acts, you can't switch post-compilation, and only one will be available anyway. That's why you do that selection so early in the build process. So if you want PDS and MuQSS, you need one kernel for each.

`Do you recommend disabling NUMA? Seems like on the forum people are enabling without issues.`
Disabling NUMA will break CUDA on Nvidia equipped systems. That also means it'll break nvenc. So if you aren't planning to use nvenc (to record with OBS for example) or CUDA, you can disable NUMA for sure.

`Is zenify for application development? I can't find much on search query.`
This is a custom patchset I made based on the Liquorix/Zen kernels. It'll improve system interactivity further than PDS or MuQSS alone (and help on the gaming front) but may slightly affect throughput on things like heavy compilation work for example. For a gaming setup, enabling it is the recommended setting.

`I selected 27 for obvious reasons but is 26 notched because it is the default selection for most compilers?`
26 is default for compilers, but also for shareability. A kernel compiled with specific optimizations can/will break on a different machine. If you're planning to use the resulting kernel only on the machine you're building it, you can safely select native or specify µarch. If your specific CPU µarch is available in the list, it is recommended to use it over native (27) btw, as the compiler might not always use the best path by itself.

`As I have hyperthreading enabled on my CPU, this is a yes, correct?`
Correct.

`Is this the same timing frequency as for the mouse polling rate?`
This defines the frequency of kernel interrupts. Various parts of the kernel are using the timer interrupt handler like scheduler time slicing, process management or simply internal time. A high value is better for interactivity (gaming for example), while a low value will have less overhead and be better for throughput. The overhead on a desktop machine is negligible though, so the throughput difference might not even be measurable.

`I had to install flex, could you add as dependency?`
Sure thing.

`cpupower.service fails to load on boot, I disabled as I don't need it. Do you encounter this?`
No problem with cpupower on my side, but I'm pretty sure it requires a module.. Can't remember which one though. If you did build with modprobed-db, it might just be missing (that definitely happened to me before). You can try to call a few cpupower functions and see if something fails. The ones requiring root privileges especially. That will probably give a hint on the issue, and if it applies, the missing module's name to add to the db.

`For the nvidia-dev-all build,I am using an PRIME Synch 90-mhwd.conf for my configuration. Should I replace some of my configuration with information from 10-nvidia-drm-outputclass.conf? Do you sometimes build the standard nvidia-all or keep it for those not looking for a vulkan release?`
I suppose this isn't required.
The nvidia-all package is following the latest short lived branch or beta if available, while the nvidia-dev-all package is following only the vulkan dev branch. The vulkan dev branch changes can sometimes get merged into the short lived/beta branches with added functionalities, or a specific fix can be added in one before the other. This is a case by case issue basically.
For example, as of right now the 396.54.06 vulkan dev driver contains a fix for some z-fighting issues on various CryEngine based games while using DXVK, while the 410.57 beta doesn't. Yet the 410.57 contains various other bugfixes regarding PRIME and wayland, as well as the new raytracing libraries. Pick your poison :D

**Comment by ryanmusante** [comment]
**I connected and disconnected my external CD-ROM, plugged and unplugged USB thumb drive, mounted an ISO, started a few obscure applications so my database is slightly larger this time around for kernel compile.

For my Nvidia PRIME sync, it has modesetting enabled. Does this message mean that since uvm,drm and modeset were not found, that they are now added? Are these messages pertaining to _localmodconfig_?
```
using config: '.config'
nvidia_uvm config not found!!
nvidia_drm config not found!!
nvidia_modeset config not found!!
```
I have never connected anything with GPIO that I'm aware of. It states on search that these are usually found on Miro, Hauppauge and STB TV-cards. Curious as to why it's asking me.
```
*
* Restart config...
*
*
* PCI GPIO expanders
*
AMD 8111 GPIO driver (GPIO_AMD8111) [N/m/y/?] n
BT8XX GPIO abuser (GPIO_BT8XX) [N/m/y/?] (NEW) 
```
Studio evolution, I answer no but will this question pop up again in the following builds or is this a once and done config? I don't like to locked down to something permanently and require manual editing.
```
  Studio Evolution SE6X (SND_SE6X) [N/m/?] (NEW) 
```
It states that this is to reduce the size of compiled kernel. Since the updated ext4 is backwards  compatible to ext2, I have this set to Yes.
```
  Use ext4 for ext2 file systems (EXT4_USE_FOR_EXT2) [Y/n/?] (NEW) 
```

**Comment by ryanmusante** [comment]
Request add as default for nvidia.
```
# For Arch Linux, some examples are given below.
IGNORE=(nvidia nvidia_drm nvidia_modeset nvidia_uvm vboxdrv vboxnetflt vboxnetadp 
vboxpci lirc_dev lirc_i2c osscore oss_hdaudio oss_usb tp_smapi thinkpad_ec zavl 
znvpair zunicode zcommon zpios zfs spl splat)
```

**Comment by Tk-Glitch** [comment]
No idea if it makes sense poking modprobed-db with that as it's configuration dependent. The messages aren't anything blocking/critical.

Perfect output for PDS :)

**Comment by ryanmusante** [comment]
Are you active on the Manjaro forums? I have asked Philip Muller about your kernels before and he is familiar with it. I opened a thread on the developer forum about adding the PDS patchset into the kernel as an option. Also I asked about adding nvidia-dev as mhwd install option.

**Comment by Tk-Glitch** [comment]
I am not but he might know me for my Android-related work, especially on the kernel side.

Nvidia-dev can move fast at times, but yeah that would definitely be nice to have that option in mhwd.

**Comment by Tk-Glitch** [comment]
The proprietary nvidia drivers are external modules, so that's the intended behavior.

**Comment by ryanmusante** [comment]
Have you experimented with LLVM 7? Do you know if there is an impact on performance compard to stable?
```
https://www.archlinux.org/packages/staging/x86_64/clang/
https://www.archlinux.org/packages/staging/x86_64/llvm/
https://www.archlinux.org/packages/staging/x86_64/llvm-libs/
```

**Comment by Tk-Glitch** [comment]
At this point, LLVM 7 is both faster and stabler than 6.

**Comment by ryanmusante** [comment]
My /etc/mkinitcpio.conf contains: 
```
MODULES="nvidia nvidia_modeset nvidia_uvm nvidia_drm"
```
I want to keep DRM kernel modesetting but I get errors now with the compile. Do you use modesetting?

**Comment by Tk-Glitch** [comment]
I'm pretty sure you need both.

**Comment by ryanmusante** [comment]
So, apparently it's not a perfect process. I found that removing the entries from mkinitcpio.conf, rebuilding nvidia driver, adding back the entries and rebuilding fixed my issue. Now I have KMS and PRIME Sync working. I'm using your "nvidia-all" branch for stability.

> Again, this sometimes breaks during driver updates. if it does, just remember to remove them from /boot/loader/entries/arch.conf and /etc/mkinitcpio.conf

https://www.gloriouseggroll.tv/arch-linux-nvidia-prime-laptop-guide-using-nvidia-package-with-fairly-new-laptops/

**Comment by ryanmusante** [comment]
Issues solved. Thread closed.
